node/src/gql/mutation/me/update-subject.ts

```diff
@@ -57,53 +100,34 @@ export const updateSubject = {
     context: { user: User }
   ): Promise<Subject> => {
     const subjectUpdate = args.subject;
-    for (const [i, questionUpdate] of (
+    const subject: any = { ...subjectUpdate };
+    subject._id = idOrNew(subjectUpdate._id);
+    subject.questions = subjectUpdate.questions || [];
+
+    for (const [i, sQuestionUpdate] of (
       subjectUpdate.questions || []
     ).entries()) {
-      for (const [i, topicUpdate] of (questionUpdate.topics || []).entries()) {
-        topicUpdate._id = idOrNew(topicUpdate._id);
-        const t = await TopicModel.findOneAndUpdate(
-          {
-            _id: topicUpdate._id,
-          },
-          {
-            $set: {
-              ...topicUpdate,
-            },
-          },
-          {
-            new: true,
-            upsert: true,
-          }
-        );
-        questionUpdate.topics[i] = t;
-      }
-      const question: any = { ...questionUpdate };
-      if (questionUpdate.topics) {
-        question._id = idOrNew(questionUpdate._id);
-        question.topics = questionUpdate.topics.map((t) => t._id);
-      }
+      const questionUpdate = sQuestionUpdate.question;
```

Comment:
OK. I understand the pattern better now. You have a {SomeType}Input graphql input object that may or may not include an `_id` property. 

If the `_id` prop is set, then you will use it for an update, otherwise, you generate a new id. Then you also use the input object as the collection of `$set` properties for mongo...

So one problem with this is it means you're passing `_id` as a property to `$set`, which we really should never do. 

What I recommend is we replace the `idOrNew` utility function with one that handles both ensuring that an `_id` exists and also strips that `_id` out of the set properties. 

I think the util function could look like this:

```typescript
interface IdAndProps<T> {
  _id: mongoose.Types.ObjectId,
  props: Partial<T>
}
export function toUpdateProps<T>(
  update: Partial<T>,
  idKeyName = "_id"
): IdAndProps<T> {
  return {
    _id: idOrNew(update._id),
   props: Object.getOwnPropertyNames(update).reduce(
     (acc: Partial<T>, cur: string) => {
       if(cur !== idKeyName) {
         acc[cur] = update[cur];
       }
       return acc;
     }, {}
  }
}
```

...and then to use it

```typescript
const {_id, props} = toUpdateProps<Question>(args.update)
```

Reply:
typescript really doesn't like that function, causes errors in all of the tests.

Reply:
let me try to add...

Reply:
https://github.com/mentorpal/mentor-graphql/pull/13